File: random_webm.py
#!/usr/bin/env python3

from sys import stdout
from urllib.request import urlopen
from urllib.parse import urljoin
import os, posixpath, tempfile
import json, re, random, argparse

parser = argparse.ArgumentParser()
parser.add_argument("-b", "--board")
parser.add_argument("-p", "--playlist", action="store_true")
args = parser.parse_args()

CHAN_HOST = "http://2ch.hk/"
board = args.board or "a"
player_string = "mpv --really-quiet "
player_playlist_options = "--shuffle --playlist="

# Internal variables
ATTACH_VIDEO = 6
catalog = urljoin(CHAN_HOST, posixpath.join(board, "catalog_num.json"))
catalog_data = json.loads(urlopen(catalog).read().decode('utf-8'))
webm_match = lambda s : re.match(r'^[^\.]*WEBM', s, re.IGNORECASE)

def get_vids(board):
    '''Retrieve list of webms on the board in {path, fullname} format'''

    webmthreads = []
    webms = []

    for thr in catalog_data['threads']:
        if webm_match(thr['subject']) or webm_match(thr['comment']):
            webmthreads.append(thr['num'])

    for thr in webmthreads:
        thr_url = urljoin(CHAN_HOST, posixpath.join(board, "res", thr + ".json"))
        thr_data = json.loads(urlopen(thr_url).read().decode('utf-8'))
        for post in thr_data['threads'][0]['posts']:
            for attach in post['files']:
                if attach['type'] == ATTACH_VIDEO and attach['nsfw'] == 0:
                    webms.append({'path': urljoin(CHAN_HOST, attach['path']), 'fullname': attach['fullname']})
    return webms

def make_playlist(vids):
    '''Make a playlist in PLS format with an iterator'''

    yield '[playlist]\n'
    for idx, webm in enumerate(vids):
        yield 'File{0}={1}\n'.format(idx, webm['path'])
        yield 'Title{0}={1}\n'.format(idx, webm['fullname'])
        yield '\n'
    yield 'NumberOfEntries={}\n'.format(len(vids))
    yield 'Version=2\n'

if __name__ == '__main__':
    pl_file = None
    vids = get_vids(board)

    if args.playlist:
        pl_file = tempfile.NamedTemporaryFile('wt')
        pl_file.writelines(make_playlist(vids))
        player_string += player_playlist_options
        destination = pl_file.name
        print('Playing all webms on /{}/...'.format(board))
    else:
        vid = random.choice(vids)
        destination = vid['path']
        print('Playing {}...'.format(vid['fullname']))

    os.system(player_string + destination)
    pl_file.close() if args.playlist else False

File: random_webm.rb
#!/usr/bin/env ruby

require 'tempfile'
require 'json'
require 'open-uri'
require 'optparse'

CHAN_HOST = "https://2ch.hk/"
board = "a"
playlist = false
player_string = "mpv --really-quiet "
player_playlist_options = "--shuffle --playlist="

ATTACH_VIDEO = 6

Webm = Struct.new(:path, :fullname)

def get_vids(board)
  catalog = File.join CHAN_HOST, board, "catalog_num.json"
  catalog_data = JSON.parse(URI.open(catalog).read)

  webm_match = ->(s) { /^[^\.]*WEBM/.match? s }
  webm_filter = ->(attach) { 
    attach["type"] == ATTACH_VIDEO && attach["nsfw"] == 0 
  }

  board_threads = catalog_data['threads'].filter do |thr| 
    webm_match.call(thr['subject']) || webm_match.call(thr['comment'])
  end 
  thread_nums = board_threads.map { |thr| thr['num'] }
  threads = thread_nums.map do |tnum|
    url = File.join CHAN_HOST, board, "res", tnum + ".json"
    tdigest = JSON.parse(URI.open(url).read)
    tdigest["threads"][0]
  end

  attaches = threads
    .map { |thr| thr['posts'] }
    .flatten
    .map { |post| post['files'] }
    .flatten

  webms = attaches.filter(&webm_filter)
  webms = webms.map do |attach| 
    Webm.new(File.join(CHAN_HOST, attach['path']), attach['fullname'])
  end

  webms
end

def make_playlist(vids)
  yield "[playlist]\n"
  vids.map.with_index do |webm, idx|
    # Make a playlist in PLS format with an iterator
    yield "File#{idx}=#{webm.path}\n"
    yield "Title#{idx}=#{webm.fullname}\n"
    yield "\n"
  end
  yield "NumberOfEntries=#{vids.length}\n"
  yield "Version=2\n"
end

OptionParser.new do |p|
  p.on("-b", "--board b") { |b| board = b }
  p.on("-p", "--playlist") { playlist = true }
end.parse!

vids = get_vids(board)

abort "No webms found on /#{board}/" if vids.empty?

destination = String.new

if playlist
  pl_file = Tempfile.new
  player_string += player_playlist_options
  destination = pl_file.path
  make_playlist(vids) { |s| pl_file.write s }
  pl_file.rewind
  puts "Playing all webms on /#{board}/..."
else
  vid = vids.sample
  destination = vid.path
  puts "Playing #{vid.fullname}..."
end

`#{player_string + destination}`
